Sweetener.Logging/LogPattern.cs

```diff
@@ -0,0 +1,385 @@
+        {
+            // TODO: We should only retrieve process and thread information if it is going
+            //       to be logged as part of the format string.
+            Process currentProcess = Process.GetCurrentProcess();
+            Thread  currentThread  = Thread.CurrentThread;
+
+            // TODO: It seems silly to use the object[] override here when we avoid the
+            //       object[] allocation in the ILogger interface. We should figure out
+            //       how to avoid this allocation too if possible.
+            return string.Format(provider, CompositeFormatString,
+                message,                       // {0} - Message
+                timestamp,                     // {1} - Timestamp
+                level,                         // {2} - LogLevel
+                currentProcess.Id,             // {3} - Process Id
+                currentProcess.ProcessName,    // {4} - Process Name
+                currentThread.ManagedThreadId, // {5} - Thread Id
+                currentThread.Name);           // {6} - Thread Name
+        }
+
+        private static string Resolve(string pattern)
+        {
+            if (pattern == null)
+                throw new ArgumentNullException(nameof(pattern));
+
+            bool hasMessage = false;
+            StringBuilder builder = new StringBuilder();
+            for (int i = 0; i < pattern.Length; i++)
+            {
+                char c = pattern[i];
+                switch (c)
+                {
+                    case '{':
+                        if (i + 1 < pattern.Length && pattern[i + 1] == '{')
+                        {
+                            // Escaped opened curly brace
+                            builder.Append(pattern, i, 2);
+                            i++;
+                        }
+                        else
+                        {
+                            PatternItem item = PatternItem.Parse(ref i, pattern);
+                            if (item.Parameter == Parameter.Message)
+                                hasMessage = true;
+
+                            builder.Append(item);
+                        }
+
+                        break;
+                    case '}':
+                        if (i + 1 < pattern.Length && pattern[i + 1] == '}')
+                        {
+                            // Escaped closed curly brace
+                            builder.Append(pattern, i, 2);
+                            i++;
+                        }
+                        else
+                        {
+                            throw new FormatException("Input string was not in a correct format");
+                        }
+
+                        break;
+                    default:
+                        builder.Append(c);
+                        break;
+                }
+            }
+
+            if (!hasMessage)
```

Comment:
Should we simply append `"{msg}"` to the string if it's missing instead?

Reply:
I prefer throwing the exception and being explicit about what is/is not required.